File: cores/typescript/src/mappers/nrom.ts
// Mapper 0 (NROM). Port of cores/c/src/mappers/nrom.c.

import { Mapper, Mirroring } from '../mapper';

export class Nrom extends Mapper {
    private _prgRom: Uint8Array;
    private _prgSize: number;
    private _chr: Uint8Array;
    private _chrSize: number;
    private _chrIsRam: boolean;
    private _mirroring: number;
    private _hasBattery: boolean;

    constructor(prg: Uint8Array, prgSize: number, chr: Uint8Array | null, chrSize: number, mirroring: number, hasBattery: boolean) {
        super();
        this.mapperNum = 0;
        this._prgSize = prgSize;
        this._prgRom = new Uint8Array(prgSize);
        if (prgSize > 0) this._prgRom.set(prg.subarray(0, prgSize));
        if (chrSize > 0) {
            this._chrSize = chrSize;
            this._chr = new Uint8Array(chrSize);
            this._chr.set(chr!.subarray(0, chrSize));
            this._chrIsRam = false;
        } else {
            this._chrSize = 8192;
            this._chr = new Uint8Array(8192);
            this._chrIsRam = true;
        }
        this._mirroring = mirroring;
        this._hasBattery = hasBattery;
    }

    private prgIndex(addr: number): number {
        const local = addr - 0x8000;
        const bank = this._prgSize;
        if (bank === 0) return 0;
        return local % bank;
    }

    readPrg(addr: number): number {
        if (addr < 0x8000) return 0x00;
        const idx = this.prgIndex(addr);
        if (idx >= this._prgSize) return 0x00;
        return this._prgRom[idx];
    }

    writePrg(addr: number, value: number): void { }

    readChr(addr: number): number {
        const sz = this._chrSize;
        if (sz === 0) return 0x00;
        return this._chr[addr % sz];
    }

    writeChr(addr: number, value: number): void {
        if (!this._chrIsRam) return;
        const sz = this._chrSize;
        if (sz === 0) return;
        this._chr[addr % sz] = value;
    }

    mirrorMode(): number { return this._mirroring; }
    chrIsRam(): boolean { return this._chrIsRam; }
    hasBattery(): boolean { return this._hasBattery; }

    saveState(buf: Uint8Array | null): number {
        const total = 7 + (this._chrIsRam ? this._chrSize : 0);
        if (buf === null) return total;
        let p = 0;
        buf[p++] = this._chrIsRam ? 1 : 0;
        buf[p++] = this._mirroring;
        buf[p++] = this._hasBattery ? 1 : 0;
        new DataView(buf.buffer).setUint32(p, this._chrSize, true); p += 4;
        if (this._chrIsRam && this._chrSize > 0) {
            buf.set(this._chr.subarray(0, this._chrSize), p);
            p += this._chrSize;
        }
        return p;
    }

    loadState(buf: Uint8Array, len: number): boolean {
        if (len < 7) return false;
        let p = 0;
        this._chrIsRam = buf[p++] !== 0;
        this._mirroring = buf[p++];
        this._hasBattery = buf[p++] !== 0;
        this._chrSize = new DataView(buf.buffer).getUint32(p, true); p += 4;
        if (this._chrIsRam && this._chrSize > 0) {
            if (len < p + this._chrSize) return false;
            this._chr.set(buf.subarray(p, p + this._chrSize));
            p += this._chrSize;
        }
        return true;
    }
}

File: cores/csharp/src/Mappers/Nrom.cs
using System.Runtime.CompilerServices;

namespace NesCore;

// Mapper 0 (NROM). Port of src/mappers/nrom.rs.
public sealed class Nrom : Mapper
{
    private byte[] _prgRom;
    private uint _prgSize;
    private byte[] _chr;
    private uint _chrSize;
    private bool _chrIsRam;
    private Mirroring _mirroring;
    private bool _hasBattery;

    public Nrom(byte[] prg, uint prgSize, byte[] chr, uint chrSize, Mirroring mirroring, bool hasBattery)
    {
        MapperNum = 0;
        _prgSize = prgSize;
        _prgRom = new byte[prgSize];
        if (prgSize > 0) System.Array.Copy(prg, _prgRom, (int)prgSize);
        if (chrSize > 0)
        {
            _chrSize = chrSize;
            _chr = new byte[chrSize];
            System.Array.Copy(chr, _chr, (int)chrSize);
            _chrIsRam = false;
        }
        else
        {
            _chrSize = 8192;
            _chr = new byte[8192];
            _chrIsRam = true;
        }
        _mirroring = mirroring;
        _hasBattery = hasBattery;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private uint PrgIndex(ushort addr)
    {
        uint local = (uint)(addr - 0x8000);
        uint bank = _prgSize;
        if (bank == 0) return 0;
        return local % bank;
    }

    public override byte ReadPrg(ushort addr)
    {
        if (addr < 0x8000) return 0x00;
        uint idx = PrgIndex(addr);
        if (idx >= _prgSize) return 0x00;
        return _prgRom[idx];
    }

    public override void WritePrg(ushort addr, byte value) { }

    public override byte ReadChr(ushort addr)
    {
        uint sz = _chrSize;
        if (sz == 0) return 0x00;
        return _chr[addr % sz];
    }

    public override void WriteChr(ushort addr, byte value)
    {
        if (!_chrIsRam) return;
        uint sz = _chrSize;
        if (sz == 0) return;
        _chr[addr % sz] = value;
    }

    public override Mirroring MirrorMode() => _mirroring;
    public override bool ChrIsRam() => _chrIsRam;
    public override bool HasBattery() => _hasBattery;

    public override uint SaveState(byte[] buf)
    {
        uint total = 0;
        // We serialize the mutable fields: chr (if RAM) + mirroring/has_battery flags.
        // Layout: [u8 chr_is_ram][u8 mirroring][u8 has_battery][u32 chr_size][chr bytes if ram]
        if (buf == null)
        {
            total = 7 + (_chrIsRam ? _chrSize : 0);
            return total;
        }
        int p = 0;
        buf[p++] = (byte)(_chrIsRam ? 1 : 0);
        buf[p++] = (byte)_mirroring;
        buf[p++] = (byte)(_hasBattery ? 1 : 0);
        System.BitConverter.GetBytes(_chrSize).CopyTo(buf, p); p += 4;
        if (_chrIsRam && _chrSize > 0)
        {
            System.Array.Copy(_chr, 0, buf, p, (int)_chrSize);
            p += (int)_chrSize;
        }
        return (uint)p;
    }

    public override bool LoadState(byte[] buf, uint len)
    {
        if (len < 7) return false;
        int p = 0;
        _chrIsRam = buf[p++] != 0;
        _mirroring = (Mirroring)buf[p++];
        _hasBattery = buf[p++] != 0;
        _chrSize = System.BitConverter.ToUInt32(buf, p); p += 4;
        if (_chrIsRam && _chrSize > 0)
        {
            if (len < (uint)(p + _chrSize)) return false;
            System.Array.Copy(buf, p, _chr, 0, (int)_chrSize);
            p += (int)_chrSize;
        }
        return true;
    }
}
